### ISSUE TITLE
QueryString doesn't appear to support nested values.

### ISSUE THREAD

**Issue by zyxxyzyxz** [livewire/livewire#1909]
### Description
You can do;
```php
public $name = 'test';
protected $queryString = [
    'name'
];
```

You can't do;
```php
public $filters = [
    'name' => 'test'
];
protected $queryString = [
    'filters.name'
];
// or
protected $queryString = [
    'filters' => [
        'name'
    ]
];
```

You can provide `filters` asa whole as a query string, but it ends up being super messy, especially in a use-case such as filters for something.

**Context**
- Livewire version: 2.3.0
- Laravel version: 8.11.2
- Browser: Chrome

**Comment by uteq** [comment]
Can you elaborate on what exactly does not work, for instance show what exactly is going wrong.

I work with nested query strings in one of my projects.
If you prefer to set the defaults for for example your 'filters' you should add that to your $filters property.
So for example:

```php
protected $queryString = [
    'filters'
];

public $filters = ['name' => 'default name', 'search'];
```

**Comment by zyxxyzyxz** [comment]
@uteq 

I was hoping it would work in a somewhat smarter way than it already currently does. Sure, you can mark the entire `filters` property as a query string, but when using this on big scale apps, this leads to absolutely ridiculous query string lengths when only 1 of them is being used.

To use your example, say somene types into the search field, but say name is blank/default. It would be ideal if it would only append the search querystring since it's the only one actually being used.

As shown in my original post, this could hopefully be solved by allowing `except` on nested queryString values, as well as being able to specify them in the first place. Because it seems at the moment you can only queryString the entire filters array which isn't too great since especially with an example like filters, you might have 30-40 options and only 1 of them is being used, so most of the query string is redudant anyway.

E.g say you set search to "test123"
```php
public $filters = [
    'search' => 'test123',
    'name' => ''
];

protected $queryString = [
    'filters.search' => ['except' => ''],
    'filters.name' => ['except' => '']
];

// Results in: https://example.org/example?search=test123
// `name` is not added to the query string because it hasn't been changed.
```

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by zyxxyzyxz** [comment]
REOPEN

**Comment by IP-Developer** [comment]
+1 for this.

**Comment by vdvcoder** [comment]
@calebporzio can you please add this feature? Thanks in advance!

**Comment by Rub-s** [comment]
+1 for this

**Comment by adesugbaa** [comment]
I have a similar issue.

```
public $filters = [
        "search" => null,
        "status" => null,
        "resolved_by" => null,
        "actively_with" => null,
        "region" => null,
        "zone" => null,
        "branch" => null,
        "vendor" => null,
        "min_created_at" => '', //'1/1/2022',
        "max_created_at" => null, //now()->format('j/n/Y'),
];

protected $queryString = [
        'sorts' => ['except' => ''],
        'filters',
        // 'filters' => [
        //     "search" => ['except' => null],
        //     "status" => ['except' => null],
        //     "resolved_by" => ['except' => null],
        //     "actively_with" => ['except' => null],
        //     "region" => ['except' => null],
        //     "zone" => ['except' => null],
        //     "branch" => ['except' => null],
        //     "vendor" => ['except' => null],
        //     "min_created_at" => '', //'1/1/2022',
        //     "max_created_at" => ['except' => null],
        // ],
    ];
```

```
<x-input.date2 wire:model="filters.min_created_at" id="filter-created-at-min" />
```

If the query string is set for any of the filters except min_created_at and max_created_at e.g. filters[region]=1, then it causes the min_created_at and max_created_at to fail when the model is bound to the trix editor.

```
Livewire Entangle Error: Livewire property 'filters.max_created_at' cannot be found
```

**Comment by vornde01** [comment]
+1 vote for this issue

**Comment by PostScripton** [comment]
Hi everyone! I figured it out.

```php
// YourComponent or trait

public array $filters = [
    'id' => '',
    'name' => '',
    'isVerified' => 'yes',
];

public $queryString = [
    'filterId' => ['except' => '', 'as' => 'id'],
    'filterName' => ['except' => '', 'as' => 'name'],
    'filterIsVerified' => ['except' => 'yes', 'as' => 'verified'],
];

public function __get($property)
{
    if (str_starts_with($property, 'filter')) {
        $key = Str::camel(Str::after($property, 'filter')); // id, name, isVerified

        return $this->filters[$key];
    }

    return parent::__get($property); // This line is important for Livewire
}
```

The URI would be like this:

```
...?id=18&name=John&verified=no
```

Of course the `filter` prefix is optional, and you can omit the if-statement in `__get()`

**Comment by livelogic** [comment]
REOPEN 
:)

Agree this is not a minor shortcoming. Without dot notation, query strings (and array props) lose so much of their potential. 
The viable strategies are: either don't use queryString at all (bad), or don't use array props for filters (better but sad).
The second option clearly denies smart approaches like the one suggested in the Surge demo:
```
public function updatedFilters()
{
    $this->resetPage();
}
```

**Comment by sathio** [comment]
@PostScripton looks like your method will not fill the form elements upon reload, am I wrong?

**Comment by PostScripton** [comment]
> @PostScripton looks like your method will not fill the form elements upon reload, am I wrong?

Dunno. Can't remember whether I've tested it or not. Though couldn't find any troubles with this solution on my production server since then. Try it our and let me know please :D Maybe I should fix it on my prod 😂

**Comment by sathio** [comment]
> > @PostScripton looks like your method will not fill the form elements upon reload, am I wrong?
> 
> 
> 
> @sathio Dunno. Can't remember whether I've tested it or not. Though couldn't find any troubles with this solution on my production server since then. Try it our and let me know please :D Maybe I should fix it on my prod 😂

yeah i think you should have a look at it ))

**Comment by PostScripton** [comment]
@sathio No, I've just checked it out and it works perfectly fine. After a reload it takes values from query string and then passes them to inputs via `wire:model`. `wire:model` works because of my `__get()` magic method. Thus, that's a working solution

**Comment by sathio** [comment]
> ```
> 
> public $queryString = [
>     'filterId' => ['except' => '', 'as' => 'id'],
>     'filterName' => ['except' => '', 'as' => 'name'],
>     'filterIsVerified' => ['except' => 'yes', 'as' => 'verified'],
> ];
> ```

weird, I can't get it to work, the querystring I get is `search?filterTxt=lol`

**Comment by sathio** [comment]
> @sathio No, I've just checked it out and it works perfectly fine. After a reload it takes values from query string and then passes them to inputs via `wire:model`. `wire:model` works because of my `__get()` magic method. Thus, that's a working solution

weird, I can't get it to work as you have mentioned, I understand how the trait should work, but when i dump $this->filters it is empty if the querystring is aliased. I wonder what I'm doing wrong
```

    public array $filters = ['txt' => ''];
    public $queryString = ['filterTxt' => ['except' => '', 'as' => 'txt']]  ;

    /search?txt=lol

array:1 [▼ // app/Http/Livewire/Traits/WithFilters.php:30
  "txt" => ""
]

```

**Comment by PostScripton** [comment]
@sathio, oh, I beg my pardon. I took a look at my enhanced version of this code and I saw that I initialize these filters values on boot.

Basicly you should do something like this:

```php
public function boot(): void
{
    $this->filters = [
        'txt' => request()->query('txt'),
    ];
}
```

Automate this thing via `$this->filters = map($this->filters)` blah blah blah...

Remember that `boot()` method in traits is written differently, so in your case it should be `bootWithFilters()`:
See [here](https://laravel-livewire.com/docs/2.x/traits#introduction) for more information.

**Comment by PostScripton** [comment]
@sathio Have you managed to make it work?
